### PR TITLE
Automatically import Azure tags

### LINKED ISSUE #15150: Server Access integration for Azure
* Teleport automatically imports node labels when running on Azure VMs
* Teleport nodes can join cluster without using tokens (analogue IAM/EC2 method)
* Teleport can automatically discover and enroll Azure VMs similar to https://github.com/gravitational/teleport/pull/12410

Fixes https://github.com/gravitational/teleport/pull/16218.

### LINKED ISSUE #15817: Automatically import Azure VM tags to servers
Similar to https://github.com/gravitational/teleport/issues/11627 we want to attach Azure VM labels to Teleport nodes. 

https://docs.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux

### PR DESCRIPTION
**Pull request by atburke**
This PR lets Teleport automatically import tags from the Azure instance it's running on.

Resolves #15817. Implements RFD 72 as of #15881.

### CODE REVIEW

**Comment by GavinFrazar** [comment]
Is this for azure servers that were proxied via static Teleport uri configuration?

**Comment by atburke** [comment]
@GavinFrazar No, this is for Teleport instances running directly on Azure (we do the same thing for EC2). I'll clarify that in the description.

File: lib/cloud/azure/imds.go
Comment (jakule) [inline lib/cloud/azure/imds.go:29]: Is there any reason why we need 1kk here, then can you move it to `constants.go` and use for both implementations, otherwise I think you can just re-use https://github.com/gravitational/teleport/blob/c219962cccdef534d640f321804874c6bc876425/constants.go#L501-L509

File: lib/cloud/azure/imds.go
Comment (jakule) [inline lib/cloud/azure/imds.go:86]: You can probably just use this https://github.com/gravitational/teleport/blob/4d25801fab927f3dc14b6eed75404da237969d66/lib/defaults/defaults.go#L776

File: lib/cloud/azure/imds_test.go
Comment (jakule) [inline lib/cloud/azure/imds_test.go:166]: nit:
```suggestion
				require.ErrorContains(t, err, tc.wantMessage)
```

File: lib/cloud/imds.go
Comment (jakule) [inline lib/cloud/imds.go:62]: 250ms sounds a bit low. Any reason why we cannot wait for at least a second or two?

File: lib/labels/cloud.go
Comment (jakule) [inline lib/labels/cloud.go:74]: Can you rename the mutex to `muLabels` so the name indicates that it protects the `labels`?

File: lib/labels/cloud.go
Comment (jakule) [inline lib/labels/cloud.go:134]: You can move the initialization right below the map is needed.

File: lib/labels/cloud_test.go
Comment (jakule) [inline lib/labels/cloud_test.go:52]: Can you add a message to the error?

**Review by jakule — COMMENTED** [review]
First pass

File: lib/cloud/imds.go
Comment (atburke) [reply lib/cloud/imds.go:62]: The tricky part is that the longer we wait here, the longer Teleport takes to start up when not running on a cloud instance. We can adjust this timeout if we want as long as we keep that tradeoff in mind (IIRC in practice a successful hit generally takes <10ms).

Comment (fspmarshall) [inline lib/cloud/imds.go:62]: A timeout as short as this seems like something that would be better implemented on a per-provider bases rather than setting it as a single top-level value.  I'd suggest also adding comments specifically justifying why such a short timeout is acceptable in this case, as super short timeouts are usually problematic.

**Review by fspmarshall — APPROVED** [review]

File: lib/cloud/aws/imds.go
Comment (r0mant) [inline lib/cloud/aws/imds.go:68]: Nit: Add godoc.

File: lib/cloud/aws/imds.go
Comment (r0mant) [inline lib/cloud/aws/imds.go:139]: Nit: Add godoc.

File: lib/cloud/azure/errors.go
Comment (r0mant) [inline lib/cloud/azure/errors.go:56]:
```suggestion
		return trace.Wrap(err)
```

File: lib/cloud/azure/imds.go
Comment (r0mant) [inline lib/cloud/azure/imds.go:65]: Nit: Add godoc.

File: lib/cloud/azure/imds.go
Comment (r0mant) [inline lib/cloud/azure/imds.go:136]: Should `apiVersion` field be protected from concurrent access?

File: lib/cloud/azure/imds.go
Comment (r0mant) [inline lib/cloud/azure/imds.go:147]:
```suggestion
		return nil, trace.NotFound("instance metadata is not available")
```

File: lib/cloud/azure/imds.go
Comment (r0mant) [inline lib/cloud/azure/imds.go:178]:
```suggestion
		return "", trace.NotFound("tag %q not found", types.CloudHostnameTag)
```

File: lib/cloud/azure/imds.go
Comment (r0mant) [inline lib/cloud/azure/imds.go:192]:
```suggestion
		return "", trace.NotImplemented("tags not supported (requires minimum API version %v, current version is %v)", minimumVersion, targetVersion)
```

File: lib/cloud/azure/imds.go
Comment (r0mant) [inline lib/cloud/azure/imds.go:187]: Can this error message be made more helpful? E.g. how/when can this happen?

Also, returned error messages should start with lowercase.

File: lib/cloud/imds.go
Comment (r0mant) [inline lib/cloud/imds.go:31]: Nice :+1:

File: lib/cloud/imds.go
Comment (r0mant) [inline lib/cloud/imds.go:62]: To add on top of other comments here - can we also move this into a constant?

File: lib/cloud/imds.go
Comment (r0mant) [inline lib/cloud/imds.go:88]:
```suggestion
		return nil, trace.NotFound("no instance metadata service found")
```

File: lib/labels/cloud.go
Comment (r0mant) [inline lib/labels/cloud.go:59]:
```suggestion
		return trace.BadParameter("missing parameter: Client")
```

File: lib/labels/cloud.go
Comment (r0mant) [inline lib/labels/cloud.go:65]: It's no longer just "ec2 labels", right? So the logger component label might need to be updated here.

File: lib/labels/cloud.go
Comment (r0mant) [inline lib/labels/cloud.go:84]: Nit: Add godoc.

File: lib/labels/cloud.go
Comment (r0mant) [inline lib/labels/cloud.go:38]: Nit: Otherwise it's not clear what "namespace" is being meant here.
```suggestion
	// AWSLabelNamespace is used as the namespace prefix for any labels
	// imported from AWS.
	AWSLabelNamespace = "aws"
	// AzureLabelNamespace is used as the namespace prefix for any labels
	// imported from Azure.
	AzureLabelNamespace = "azure"
```

**Review by r0mant — APPROVED** [review]
lgtm with a few nits, but please also wait for @jakule approval.

**Review by jakule — APPROVED** [review]

**Pull request merged**
